scipy/stats/_qmc.py

```diff
@@ -481,6 +485,186 @@ def _perturb_discrepancy(sample: np.ndarray, i1: int, i2: int, k: int,
+    sample : array_like (n, d)
+        The sample after an iteration of Lloyd's algorithm.
+
+    """
+    sample = _jitter_sample(sample, rng=rng)
+    new_sample = np.empty_like(sample)
+
+    voronoi = spatial.Voronoi(sample)
+
+    for ii, idx in enumerate(voronoi.point_region):
+        # the region is a series of indices into self.voronoi.vertices
+        # remove point at infinity, designated by index -1
+        region = [i for i in voronoi.regions[idx] if i != -1]
+
+        # get the vertices for this region
+        verts = voronoi.vertices[region]
+
+        # clipping would be wrong, we need to intersect
+        # verts = np.clip(verts, 0, 1)
+
+        # move sample towards centroids:
+        # Centroid in n-D is the mean for uniformly distributed nodes
+        # of a geometry.
+        centroid = np.mean(verts, axis=0)
+        new_sample[ii] = sample[ii] + (centroid - sample[ii]) * decay
+
+    # only update sample to centroid within the region
+    is_valid = np.all(np.logical_and(new_sample >= 0, new_sample <= 1), axis=1)
+    sample[is_valid] = new_sample[is_valid]
+
+    return sample
+
+
+def lloyd_centroidal_voronoi_tessellation(
+        sample: npt.ArrayLike,
+        *,
+        n_iters: IntNumber = 10,
+        seed: SeedType = None
+) -> np.ndarray:
+    """Centroidal Voronoi Tessellation.
+
+    Perturb a sample of points in :math:`[0, 1]^d` using Lloyd's algorithm.
+
+    Parameters
+    ----------
+    sample : array_like (n, d)
+        The sample to iterate on. With ``n`` the number of sample and ``d``
+        the dimension.
+    n_iters : int, optional
+        Number of iterations. Too many iterations tend to cluster the sample
+        as a hypersphere. Default is 10.
+    seed : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` instance then that instance is
+        used.
+
+    Returns
+    -------
+    sample : array_like (n, d)
+        The sample after `n_iters` of Lloyd's algorithm.
+
+    Notes
+    -----
+    Lloyd's algorithm is an iterative process which purpose is to improve
+    the dispersion of a sample. For a given sample: (i) compute a Voronoi
+    Tessellation; (ii) find the centroid of each Voronoi cell; (iii) move the
+    samples toward the centroid of their respective cell.
+
+    The process converges to equally spaced points. It implies that measures
+    like the discrepancy could suffer from too many iterations. On the other
+    hand, L1 and L2 distances should improve. This is especially true with
+    QMC methods which tend to favor the discrepancy over other criteria.
+
+    .. warning::
+
+       The Voronoi Tessellation step is expensive and quickly becomes
+       intractable with dimensions as low as 10 even for a sample of size
+       as low as 1000.
```

Comment:
As you likely know, the dimension problem is most unfortunate--I believe machine learning applications tend to have lots of dimensions, which makes this kind of thing a bit hard to use, but not much we can do about that.

One idea might be to leave a back door open for future expansion to approximate methods that don't use exact Voronoi diagrams. I admit I didn't come up with that idea myself--it is on Wikipedia right here: https://en.wikipedia.org/wiki/Lloyd%27s_algorithm#Approximation

Mind bendingly:
- Monte Carlo methods can be used to avoid the exact Voronoi tesselation (and we're already in QMC here, but at least if we moved to `spatial` it may be that the user is like me and barely knows what qmc is, so the input data may not have been treated that way yet)
- there's also a mention of using pixel grids/graphics strategies to avoid exact tesselations (might be more @stefanv 's domain than mine)

Anyway, I'm just thinking about leaving a back door open, not asking you to implement that stuff at the moment.

Reply:
Totally agree, here there is not much we can do. Still it's a valuable method for low dimensional problems.

What do you have in mind when you say that we should leave a back door open? I mean concretely in the code, what do you suggest we do?

Do we have anything for non exact Voronoi tessellation? Or do you say we could have this at one point and make sure we could pass a param to do this?

Reply:
Not sure there is too much we have to do right now, can always add a parameter later to change the backend algorithm I think? From the user perspective the only requirement is to get the same shape of data back out as they put in I think, so probably gives a fair bit of design flexibility already on what can happen inside the function.

Reply:
I agree that a parameter could be added for approximate tessellation in the future.